src/piece.js

```diff
@@ -1,140 +1,188 @@
  */
-export const fromJSON = ({ link, height }) =>
-  new PieceInfo({ link: Link.parse(link['/']), height })
+export const toJSON = (piece) => ({ '/': toString(piece) })
 
 /**
- * @param {API.PieceInfo} piece
- * @returns {API.ToString<API.PieceJSON>}
+ * @param {Uint8Array} payload
  */
-export const toString = (piece) => JSON.stringify(toJSON(piece), null, 2)
+export const fromPayload = (payload) => fromDigest(digest(payload))
 
 /**
- * @param {API.ToString<API.PieceJSON>|string} source
+ * @param {API.Piece} piece
+ * @returns {API.PieceView}
  */
-export const fromString = (source) => fromJSON(JSON.parse(source))
+export const toView = (piece) => fromDigest(toDigest(piece))
 
 /**
- * Creates Piece CID from the the merkle tree root. It will not perform
- * any validation.
  *
- * @param {API.MerkleTreeNode} root
+ * @param {API.Piece} piece
  * @returns {API.PieceLink}
  */
-export const createLink = (root) =>
-  Link.create(
-    FilCommitmentUnsealed,
-    Digest.create(SHA2_256_TRUNC254_PADDED, root)
-  )
+export const toLink = (piece) => Link.create(Raw.code, toDigest(piece))
 
 /**
- * @param {Uint8Array} source
- * @returns {API.Piece}
+ *
+ * @param {API.Piece} piece
+ * @returns {API.PieceInfo}
+ */
+export const toInfo = (piece) => new Info(toDigest(piece))
+
+/**
+ *
+ * @param {API.PieceInfo} info
+ * @returns
  */
-export const build = (source) => {
-  if (source.length < MIN_PAYLOAD_SIZE) {
-    throw new RangeError(
-      `Piece is not defined for payloads smaller than ${MIN_PAYLOAD_SIZE} bytes`
-    )
+export const fromInfo = (info) =>
+  toView({
+    height: PaddedSize.toHeight(info.size),
+    root: info.link.multihash.digest,
+  })
+
+class Piece {
+  /**
+   * @param {API.PieceLink} link
+   */
+  constructor(link) {
+    this.link = link
+  }
+  get height() {
+    return this.link.multihash.bytes[PREFIX.length]
+  }
+  get size() {
+    return PaddedSize.fromHeight(this.height)
+  }
+  get root() {
+    return this.link.multihash.bytes.subarray(PREFIX.length + 1)
   }
 
-  if (source.length > MAX_PAYLOAD_SIZE) {
-    throw new RangeError(
-      `Payload exceeds maximum supported size of ${MAX_PAYLOAD_SIZE} bytes`
-    )
+  toJSON() {
+    return {
+      '/': this.toString(),
+    }
+  }
+  toString() {
+    return /** @type {API.ToString<API.PieceLink>} */ (this.link.toString())
   }
 
-  const tree = Tree.build(Fr32.pad(source))
+  toInfo() {
+    return new Info(this)
+  }
+}
 
-  return new Piece({ tree, contentSize: source.byteLength })
+/**
+ * @implements {API.PieceInfo}
+ */
+class Info {
+  /**
+   * @param {API.Piece} piece
+   */
+  constructor(piece) {
+    this.piece = piece
+    /** @type {API.LegacyPieceLink|undefined} */
+    this._link
+  }
+  get height() {
+    return this.piece.height
+  }
+  get root() {
+    return this.piece.root
+  }
+  get size() {
+    return PaddedSize.fromHeight(this.height)
+  }
+
+  get link() {
+    if (this._link == null) {
+      this._link = Link.create(
```

Comment:
@Gozala this creates a LegacyLink for the `AggregateBuild.toInfo()` which should not be expect right?

Reply:
I'm not sure I understand the question here. `PieceInfo` corresponds to same named [go struct](https://github.com/filecoin-project/go-state-types/blob/6de09269e66d209acc7eb38d32398710b883dd51/abi/piece.go#L47-L50) which uses older version of CID, which is why we create a legacy link here.

Reply:
It is related to getting an Aggregate CID being Piece CIDv2. I think this is important for us for metrics at least, be able to know by Aggregate CID their encoded height/size

Reply:
Now, it looks like only CID we can get from here is through `AggregateBuild.toInfo()`. Maybe we just want a getter for Piece CIDv2 from aggregate

Reply:
See what I mean in practise https://github.com/web3-storage/w3up/pull/850/files